use edit distance feature

tpeng opened the pull request:
webpager sometimes give false positive for some links. e.g. on http://www.zoeken.nu/115755/bedrijf/rabobank-drechtsteden.html. it labeled following links as pagination:
`http://www.zoeken.nu/zoek/j.html` , `http://www.zoeken.nu/zoek/t.html`, `http://www.zoeken.nu/zoek/z.html`
it might because these link has single text which has bigger weight in the logistic classifier...
but apparently they're not. this change add a edit distance between the base url (the url of the page to classify) and link's our url as one feature, this should be able to catch the idea those urls tend to be similar.

also some minor improvements.

tpeng committed:
use edit distance feature

kmike commented:
I'd prefer using some levenshtein library from pypi - there must be some :)

kmike commented on webpager/features.py:
```diff
@@ -52,18 +57,21 @@ def clean_html(cls, html, encoding=None):
         return self.clean_html(html, encoding)
 
-    def fit_transform(self, X, y=None, encoding=None):
+    def fit_transform(self, X, baseurl, y=None, encoding=None):
```
If we want to maintain consistency with sklearn it is better to rename this method as it doesn't follow the API.

kmike commented on webpager/features.py:
```diff
@@ -23,10 +26,12 @@
     return text.split()
 
-def get_anchor_text(anchor):
+def get_text(x):
```
I kind of dislike this interface - if x is a tuple `(a_elem, page_url)` why not pass 2 arguments to feature functions?

kmike commented on webpager/functions.py:
```diff
@@ -27,3 +31,9 @@ def number_pattern(anchor):
+    href = anchor.get('href', '')
+    d = levenshtein_distance(baseurl, href)
+    # normalize
+    return float(d) / max(len(baseurl), len(href))
```
Is it good to normalize the edit distance? E.g. why do we want to penalize pagination link from short domains more?

tpeng replied:
good catch! now the distance is calculate base on the path + params + QS + fragment (i.e. urlparse(url)[2:])

kmike commented:
I like the idea of using features calculated based on url differences, and Levenshtein distance is a reasonable approximation. More specific heuristics could help - e.g. a link to the page is likely changed in the end, not in the middle, so if URLs differ in the middle it could be a sign it is not a pagination link; Levenshtein distance won't detect that.

We don't have enough data to check if it is a good feature or not - e.g. on some websites URL of the start page can be without page number, but paginator can lead to pages with numbers - but you knows that better; feel free to introduce this feature if it helps.

rmax commented:
> e.g. a link to the page is likely changed in the end, not in the middle, so if URLs differ in the middle it could be a sign it is not a pagination link

That might be true for path-based links, like `/category/foo/page/10/`, but I've seen urls with long list of parameters in the query string having the pagination parameter at the begining of the QS. Not exactly like the url below but similar:

```
https://www.google.com.bo/search?start=20&q=github+rebase+a+pull+request&oq=github+rebase+a+pull+request&aqs=chrome..69i57.2946j0j1&client=ubuntu-browser&sourceid=chrome&ie=UTF-8#q=how+do+i+follow+links+in+scrapy+&safe=off
```

Perhaps considering just the edit distance in the QS and the last half of the path might cover both cases.

kmike commented:
A good idea about using separate distance metrics for path and query string. For query string we may also consider Jaccard similarity or other set distance, because QS may be unordered.

kmike commented:
By the way, GET argument names could be also good features - if there is "page=2" argument in a href it is more likely to be a pagination link.

rmax commented:
I haven't seen the source code in deep, but does it consider the pattern of many links in a row?

It's widely used this form of pagination:

```
<a ...>1</a> <a ...>2</a> <span>3</span> <a> 4 </a>
```

Optionally with next/prev or first/last links. So, a list of links with highly similar url can be seen as pagination links.

But, a counterexample is having a list of page links where only varies an id number:

```
Jump to page: <a href="book.php?p=1">1</a> <a href="book.php?p=2">2</a> ...
```

kmike commented:
I'm also just took a first look at webpager; it seems it doesn't use information about `<a>` nearby tags.

tpeng commented:
Thanks @kmike and @darkrho , thanks for all good suggestions! 

I added the feature to catch the "page=2" in the QS. seems like it's a good idea: these features are listed in top20 positive features!
currently we don't look around the nearby tags. but this is a good idea. some model like maximum entropy should be able to catch it.
different weights for different components (QS, path, etc) could help too, but currently i prefer the simplest way to do it: just discard the scheme and netloc and get a edit distance base on the rest.
so far, this is just a very simple linear model which seems give good performance on my testing data, maybe we can use these more sophisticated features if we find simple model not works well.

tpeng commented:
@kmike, for the 3rd levenshtein library, i prefer to use our own implementation as it has least dependency.

tpeng committed:
feature function cleanups
improve edit distance feature to use discard the scheme and netloc before comparing
add feature to catch the idea of page=2 usually show in query parameters
